Significant language and style improvement (#17389)
mdn/content: files/en-us/learn/javascript/asynchronous/introducing/index.md

{{LearnSidebar}}{{NextMenu("Learn/JavaScript/Asynchronous/Promises", "Learn/JavaScript/Asynchronous")}}

> In this article, we explain what asynchronous programming is, why we need it, and briefly discuss some of the ways asynchronous functions have historically been implemented in JavaScript.

Author: Yes, of course, I agree. The future tense is a style issue and I should have left it as it was.

  </tbody>
</table>

Asynchronous programming is a technique that enables your program to start a potentially long-running task and still be able to be responsive to other events while that task runs, rather than having to wait until that task has finished. Once that task has finished, your program is presented with the result.

> Many functions provided by browsers, and especially the most interesting ones, can potentially take a long time, and are therefore asynchronous. For example:

Author: Yes, I missed that. That's better.

// "Hello, my name is Miriam!"
```

> Here, `makeGreeting()` is a **synchronous function**, because the caller has to wait for the function to finish its work and return before the caller can continue.

Author: I am wondering about the following:

to finish its work and return a value before the caller can continue.

### A long-running synchronous function

What if the synchronous function takes a long time?

> This program generates a number of large prime numbers, using a very inefficient algorithm, when the user clicks the "Generate primes" button. The user can set the number of primes to generate, which of course affects how long the operation will take.

Reviewer: ```suggestion
The program below uses a very inefficient algorithm to generate multiple large prime numbers when a user clicks the "Generate primes" button. The user can control the number of primes to generate, which of course, affects how long the operation will take.
```
- Replaced one of the two 'number' with 'multiple' to avoid confusion
- I don't see a problem with using the word 'control'. It is not incorrect.

Author: I was thinking about controlling a program while it was executing (with a slider, for example).

I agree, control is not incorrect.

Reviewer: I see what you're saying with the use of the word 'control', but I guess it is apparent from the code example that there is no slider. How about replacing that sentence with:
"The higher the number a user specifies, the longer the operation will take."

Author: Yes, that is more simple and makes much more sense. I added "of primes" in the new commit as that is the title of the input field, giving:

The higher the number of primes a user specifies, the longer the operation will take.

{{EmbedLiveSample("A long-running synchronous function", 600, 120)}}

> Click "Generate primes". Depending on how fast your computer is, it will probably take a few seconds before the program displays the "Finished!" message.

Reviewer: ```suggestion
Try clicking "Generate primes". Depending on how fast your computer is, it will probably take a few seconds before the program displays the "Finished!" message.
```
The original is not incorrect, just a matter of style

### The trouble with long-running synchronous functions

> The next example is just like the last one, except we added a text box for you to type in. This time, click "Generate primes" and then try typing in the text box.

Reviewer: ```suggestion
The next example is just like the last one, except we added a text box for you to type in. This time, click "Generate primes", and try typing in the text box immediately after.
```

Author: Yes, much clearer.

> You'll find that while our `generatePrimes()` function is running, our program is completely unresponsive: you can't type anything, or click anything, or do anything else.

Reviewer: ```suggestion
You'll find that while our `generatePrimes()` function is running, our program is completely unresponsive: you can't type anything, click anything, or do anything else.
```

{{EmbedLiveSample("The trouble with long-running synchronous functions", 600, 200)}}

This is the basic problem with long-running synchronous functions. What we need is a way for our program to:

> - start a long-running operation by calling a function

Reviewer: ```suggestion
1. Start a long-running operation by calling a function.
```
these points are linked to each other in a sequential order, so changing them to numbered list

> - have the function start the operation and return immediately, so our program can still be responsive to other events
> - be notified with the result of the operation, when it eventually completes.

Reviewer: ```suggestion
2. Have that function start the operation and return immediately, so that our program can still be responsive to other events.
3. Notify us with the result of the operation when it eventually completes.
```

> That's what asynchronous functions can do, and the rest of this module explains how they are implemented in JavaScript.

Reviewer: ```suggestion
That's precisely what asynchronous functions can do. The rest of this module explains how they are implemented in JavaScript.
```

## Event handlers

> The description we just saw of asynchronous functions might remind you of event handlers, and if it does, you'd be right. Event handlers are really a form of asynchronous programming: you provide a function (the event handler) that will be called, not immediately, but whenever the event happens. If "the event" is "the asynchronous operation has completed", then that event could be used to notify the caller about the result of an asynchronous function call.

Reviewer: ```suggestion
The description we just saw of asynchronous functions might remind you of event handlers, and if it does, you'd be right. Event handlers are really a form of asynchronous programming: you provide a function (the event handler) that will be called, not right away, but whenever the event happens. If "the event" is "the asynchronous operation has completed", then that event could be used to notify the caller about the result of an asynchronous function call.
```
Nothing incorrect with 'right away'

> Some early asynchronous APIs used events in this way. The {{domxref("XMLHttpRequest")}} API enables you to make HTTP requests to a remote server using JavaScript. Since this can take a long time, it's an asynchronous API, and you get notified about the progress and eventual completion of a request by attaching event listeners to the `XMLHttpRequest` object.

Reviewer: ```suggestion
Some early asynchronous APIs used events in just this way. The {{domxref("XMLHttpRequest")}} API enables you to make HTTP requests to a remote server using JavaScript. Since this can take a long time, it's an asynchronous API, and you get notified about the progress and eventual completion of a request by attaching event listeners to the `XMLHttpRequest` object.
```
The original sentence is better with 'just' for emphasis.
